src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+     * @return string[] $words
+     */
+    private function prepCorpus(array $sentences) : array
+    {
+        $words = [];
+
+        foreach ($sentences as $sentence) {
+            $preppedSentence = $this->prepSentence($sentence);
+
+            $this->corpus[] = $preppedSentence;
+            $words = array_merge($words, $preppedSentence);
+        }
+
+        return $words;
+    }
+
+    /**
+     * Sanitizing & exploding a provided sentence.
+     *
+     * @param string $sentence
+     * @return string[] $exploded
+     */
+    private function prepSentence(string $sentence) : array
+    {
+        $sentence = (string) preg_replace('#[[:punct:]]#', '', strtolower($sentence));
+        $sentence = (string) preg_replace('/\s\s+/', ' ', str_replace("\n", ' ', $sentence));
+        $sentence = trim($sentence);
+        return explode(' ', $sentence);
+    }
+
+    /**
+     * Preparing and setting an initial raw vocab array for additional preprocessing
+     *
+     * @param string[] $words
+     */
+    private function scanVocab(array $words) : void
+    {
+        foreach ($words as $word) {
+            $this->rawVocab[$word] = (int) ($this->rawVocab[$word] ?? 0) + 1;
+        }
+    }
+
+    /**
+     * Iterating through each word in raw vocab and formally appending word, and corresponding word count, index, and word, to vocab property.
+     * Creating initial index2word property to manage word counts.
+     */
+    private function prepVocab() : void
+    {
+        $dropTotal = $dropUnique = $retainTotal = 0;
+        $retainWords = [];
+
+        foreach ($this->rawVocab as $word => $v) {
+            if ($v >= $this->minCount) {
+                $retainWords[] = $word;
+                $retainTotal += $v;
+
+                $this->vocab[$word] = ['count' => $v, 'index' => count($this->index2word), 'word' => $word];
+                $this->index2word[] = $word;
+            } else {
+                ++$dropUnique;
+                $dropTotal += $v;
+            }
+        }
+
+        $originalUniqueTotal = count($retainWords) + $dropUnique;
+        $retainUniquePct = ((count($retainWords) * 100) / $originalUniqueTotal);
+
+        $originalTotal = $retainTotal + $dropTotal;
+        $retainPct = (($retainTotal * 100) / $originalTotal);
+
+        $thresholdCount = $this->thresholdCount($retainTotal);
+
+        $this->updateWordProbs($retainWords, $thresholdCount);
+    }
+
+    /**
+     * Determine threshold word count based off of subsampling rate.
+     *
+     * @param int $retainTotal
+     * @return float $thresholdCount
+     */
+    private function thresholdCount(int $retainTotal) : float
+    {
+        if (!$this->sampleRate) {
+            $thresholdCount = $retainTotal;
+        } elseif ($this->sampleRate < 1) {
+            $thresholdCount = $this->sampleRate * $retainTotal;
+        } else {
+            $thresholdCount = ($this->sampleRate * (3 + sqrt(5)) / 2);
+        }
+
+        return $thresholdCount;
```

Comment:
Looks like we can return early from the if statement again ... unless you feel we need the expressiveness of this variable (I do not).